src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs

```diff
@@ -50,31 +62,55 @@ private SwaggerObjectBase Build(JToken token)
                     }
 
-                    var formatted = RestApiHelper.FormatReferenceFullPath((string)referenceToken);
-                    var deferredObject = new SwaggerReferenceObject
+                    var swaggerReference = RestApiHelper.FormatReferenceFullPath((string)referenceToken);
+                    switch (swaggerReference.Type)
                     {
-                        DeferredReference = formatted.Item1,
-                        ReferenceName = formatted.Item2,
-                        Location = location
-                    };
-
-                    // For swagger, other properties are still allowed besides $ref, e.g.
-                    // "schema": {
-                    //   "$ref": "#/definitions/foo"
-                    //   "example": { }
-                    // }
-                    // Use Token property to keep other properties
-                    // These properties cannot be referenced
-                    jObject.Remove("$ref");
-                    deferredObject.Token = jObject;
-                    _documentObjectCache.Add(location, deferredObject);
-                    return deferredObject;
+                        case RestApiHelper.SwaggerReferenceType.InternalReference:
+                            var deferredObject = new SwaggerReferenceObject
+                            {
+                                DeferredReference = swaggerReference.Path,
+                                ReferenceName = swaggerReference.Name,
+                                Location = location
+                            };
+
+                            // For swagger, other properties are still allowed besides $ref, e.g.
+                            // "schema": {
+                            //   "$ref": "#/definitions/foo"
+                            //   "example": { }
+                            // }
+                            // Use Token property to keep other properties
+                            // These properties cannot be referenced
+                            jObject.Remove("$ref");
+                            deferredObject.Token = jObject;
+                            _documentObjectCache.Add(location, deferredObject);
+                            return deferredObject;
+                        case RestApiHelper.SwaggerReferenceType.ExternalReference:
+                            jObject.Remove("$ref");
+
+                            // Currently we support external reference without $ref inside
+                            var externalSwaggerPath = Path.Combine(swaggerDir, swaggerReference.Path);
+                            using (JsonReader reader = new JsonTextReader(File.OpenText(externalSwaggerPath)))
+                            {
+                                var externalJObject = JObject.Load(reader);
+                                foreach (var item in externalJObject)
+                                {
+                                    jObject[item.Key] = item.Value;
```

Comment:
In my understanding exsiting property could be overwritten, so should warning here

---
In reply to: [93157759](https://github.com/dotnet/docfx/pull/1019#discussion_r93157759) [](ancestors = 93157759)